packages/csvviewer-extension/src/index.ts

```diff
@@ -2,74 +2,112 @@
 // Distributed under the terms of the Modified BSD License.
 
 import {
-  ILayoutRestorer, JupyterLab, JupyterLabPlugin
+    ILayoutRestorer, JupyterLab, JupyterLabPlugin
 } from '@jupyterlab/application';
 
 import {
-  InstanceTracker
+    InstanceTracker
 } from '@jupyterlab/apputils';
 
 import {
-  CSVViewer, CSVViewerFactory
+    CSVViewer, CSVViewerFactory, TSVViewerFactory
 } from '@jupyterlab/csvviewer';
 
 import {
-  IDocumentWidget
+    IDocumentWidget
 } from '@jupyterlab/docregistry';
 
 /**
- * The name of the factory that creates CSV widgets.
+ * The name of the factories that creates widgets.
  */
-const FACTORY = 'Table';
+const FACTORY_CSV = 'CSVTable';
+const FACTORY_TSV = 'TSVTable';
 
 
 /**
- * The table file handler extension.
+ * The CSV file handler extension.
  */
-const plugin: JupyterLabPlugin<void> = {
-  activate,
-  id: '@jupyterlab/csvviewer-extension:plugin',
-  requires: [ILayoutRestorer],
-  autoStart: true
+
+const csv: JupyterLabPlugin<void> = {
+    activate: (app: JupyterLab, restorer: ILayoutRestorer): void => {
+        const factory = new CSVViewerFactory({
+            name: FACTORY_CSV,
+            fileTypes: ['csv'],
+            defaultFor: ['csv'],
+            readOnly: true
+        });
+        const tracker = new InstanceTracker<IDocumentWidget<CSVViewer>>({ namespace: 'csvviewer' });
+
+        // Handle state restoration.
+        restorer.restore(tracker, {
+            command: 'docmanager:open',
+            args: widget => ({ path: widget.context.path, factory: FACTORY_CSV }),
+            name: widget => widget.context.path
+        });
+
+        app.docRegistry.addWidgetFactory(factory);
+        let ft = app.docRegistry.getFileType('csv');
+        factory.widgetCreated.connect((sender, widget) => {
+            // Track the widget.
+            tracker.add(widget);
+            // Notify the instance tracker if restore data needs to update.
+            widget.context.pathChanged.connect(() => { tracker.save(widget); });
+
+            if (ft) {
+                widget.title.iconClass = ft.iconClass;
+                widget.title.iconLabel = ft.iconLabel;
+            }
+        });
+    },
+    id: '@jupyterlab/csvviewer-extension:csv',
+    requires: [ILayoutRestorer],
+    autoStart: true
 };
 
 
+
 /**
- * Export the plugin as default.
+ * The TSV file handler extension.
  */
-export default plugin;
+const tsv: JupyterLabPlugin<void> = {
+    activate: (app: JupyterLab, restorer: ILayoutRestorer): void => {
```

Comment:
Rather than move these to inline functions, can you create two module-level functions `activateCSV` and `activateTSV`? IMHO that would be cleaner.

Reply:
Ok, thanks for suggestion, I will do it.